### PR TITLE
Create count_of_set_bits.cpp

### PR DESCRIPTION
**Pull request by Swastyy**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1515"></a>

### CODE REVIEW

**Comment by Swastyy** [comment]
@Panquesito7 Have a look into this PR 😄

**Comment by Swastyy** [comment]
@fhlasek Please have a look in this PR also. 😃

**Review by fhlasek — APPROVED** [review]
Well written, well tested, well documented. Thanks!

File: bit_manipulation/count_of_set_bits.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
+/**
+ * @file
+ * @brief Program to Count set bits in an integer
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:3]:
```suggestion
 * @brief Implementation to [count sets bits](https://www.geeksforgeeks.org/count-set-bits-in-an-integer/) in an integer.
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ * @author [Swastika Gupta](https://github.com/swastyy)
+ */
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for io operations
+#include <vector>    /// for std::vector
+
+/**
+ * @namespace bit_manipulation
+ * @brief bit set counting algorithm
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:17]:
```suggestion
 * @brief Bit manipulation algorithms
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+ */
+namespace bit_manipulation {
+/**
+ * @namespace bitCount
+ * @brief Functions for counting set bits in binary representation of a number
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:22]:
```suggestion
 * @brief Functions for the [count sets bits](https://www.geeksforgeeks.org/count-set-bits-in-an-integer/) implementation
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ */
+namespace bitCount {
+/**
+ * @brief The main function implements set bit count
+ * @param n is the number whose set bit will be counted
+ * @returns count , the number set bit in binary representation of n
+ */
+
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:30]:
```suggestion
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    assert(bit_manipulation::bitCount::countSetBits(25) == 3);
+    std::cout << "All test cases passed" << std::endl;
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();  // execute the test
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:68]:
```suggestion
    test();  // run self-test implementations
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        count += n & 1;
+        n = n >> 1;  // n=n/2
+    }
+    return count;
+}
+}  // namespace bitCount
+}  // namespace bit_manipulation
+
+/**
+ * @brief Test implementations
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:44]:
```suggestion
 * @brief Self-test implementations
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+namespace bit_manipulation {
+/**
+ * @namespace bitCount
+ * @brief Functions for counting set bits in binary representation of a number
+ */
+namespace bitCount {
+/**
+ * @brief The main function implements set bit count
+ * @param n is the number whose set bit will be counted
+ * @returns count , the number set bit in binary representation of n
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:28]:
```suggestion
 * @returns the count of the number set bit in the binary representation of `n`
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 8]
```diff
@@ -0,0 +1,70 @@
+/**
+ * @file
+ * @brief Program to Count set bits in an integer
+ *
+ * @details
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:5]:
```suggestion
 * @details
 * We are given an integer number. Let’s say, number. The task is to first calculate the
 * binary digit of a number and then calculate the total set bits of a number.
 *
 * Set bits in a binary number is represented by 1. Whenever we calculate the binary number of an integer value
 * it is formed as the combination of 0’s and 1’s. So digit 1 is known as a set bit in computer terms.
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good work!

**Comment by Panquesito7** [comment]
Sorry for taking long in reviewing this PR (including other PRs). There have been many PRs recently (plus other important stuff I have to work on). 😅

File: bit_manipulation/count_of_set_bits.cpp (hunk 8 above)
Comment (Panquesito7) [reply bit_manipulation/count_of_set_bits.cpp:7]: If you're not going to add this, please add a detailed description of the algorithm. I'm not trying to be rude, by the way. :)

File: bit_manipulation/count_of_set_bits.cpp [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+#include <iostream>  /// for io operations
+#include <vector>    /// for std::vector
+
+/**
+ * @namespace bit_manipulation
+ * @brief Bit manipulation algorithms
+ */
+namespace bit_manipulation {
+/**
+ * @namespace bitCount
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:23]:
```suggestion
 * @namespace count_of_set_bits
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+ * @brief Functions for the [count sets
+ * bits](https://www.geeksforgeeks.org/count-set-bits-in-an-integer/)
+ * implementation
+ */
+namespace bitCount {
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:28]:
```suggestion
namespace count_of_set_bits {
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+std::uint64_t countSetBits(int n) {
+    int count = 0;  // "count" variable is used to count number of 1's in binary
+                    // representation of the number
+    while (n != 0) {
+        count += n & 1;
+        n = n >> 1;  // n=n/2
+    }
+    return count;
+}
+}  // namespace bitCount
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:43]:
```suggestion
}  // namespace count_of_set_bits
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+}  // namespace bit_manipulation
+
+/**
+ * @brief Test implementations
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:47]:
```suggestion
 * @brief Self-test implementations
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+ * @returns void
+ */
+static void test() {
+    // n = 4 return 1
+    assert(bit_manipulation::bitCount::countSetBits(4) == 1);
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:52]:
```suggestion
    assert(bit_manipulation::count_of_set_bits::countSetBits(4) == 1);
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    // n = 6 return 2
+    assert(bit_manipulation::bitCount::countSetBits(6) == 2);
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:54]:
```suggestion
    assert(bit_manipulation::count_of_set_bits::countSetBits(6) == 2);
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    // n = 13 return 3
+    assert(bit_manipulation::bitCount::countSetBits(13) == 3);
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:56]:
```suggestion
    assert(bit_manipulation::count_of_set_bits::countSetBits(13) == 3);
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    // n = 9 return 2
+    assert(bit_manipulation::bitCount::countSetBits(9) == 2);
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:58]:
```suggestion
    assert(bit_manipulation::count_of_set_bits::countSetBits(9) == 2);
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+    // n = 15 return 4
+    assert(bit_manipulation::bitCount::countSetBits(15) == 4);
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:60]:
```suggestion
    assert(bit_manipulation::count_of_set_bits::countSetBits(15) == 4);
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+    // n = 25 return 3
+    assert(bit_manipulation::bitCount::countSetBits(25) == 3);
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:62]:
```suggestion
    assert(bit_manipulation::count_of_set_bits::countSetBits(25) == 3);
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+    std::cout << "All test cases passed" << std::endl;
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:63]:
```suggestion
    std::cout << "All test cases successfully passed!" << std::endl;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: bit_manipulation/count_of_set_bits.cpp [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+ * @details
+ * We are given an integer number. Let’s say, number. The task is to first
+ * calculate the binary digit of a number and then calculate the total set bits
+ * of a number.
+ *
+ * Set bits in a binary number is represented by 1. Whenever we calculate the
+ * binary number of an integer value it is formed as the combination of 0’s and
+ * 1’s. So digit 1 is known as a set bit in computer terms. Time Complexity :-
+ * O(log n) Space complexity :- O(1)
+ * @author [Swastika Gupta](https://github.com/swastyy)
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:16]:
```suggestion
 * @author [Swastika Gupta](https://github.com/Swastyy)
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+ *
+ * @details
+ * We are given an integer number. Let’s say, number. The task is to first
+ * calculate the binary digit of a number and then calculate the total set bits
+ * of a number.
+ *
+ * Set bits in a binary number is represented by 1. Whenever we calculate the
+ * binary number of an integer value it is formed as the combination of 0’s and
+ * 1’s. So digit 1 is known as a set bit in computer terms. Time Complexity :-
+ * O(log n) Space complexity :- O(1)
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:15]:
```suggestion
 * 1’s. So digit 1 is known as a set bit in computer terms.
 * Time Complexity: O(log n)
 * Space complexity: O(1)
```

File: bit_manipulation/count_of_set_bits.cpp [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+        count += n & 1;
+        n = n >> 1;  // n=n/2
+    }
+    return count;
+}
+}  // namespace count_of_set_bits
+}  // namespace bit_manipulation
+
+/**
+ * @brief Test implementations
```
Comment (Panquesito7) [inline bit_manipulation/count_of_set_bits.cpp:54]:
```suggestion
 * @brief Self-test implementations
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Review by Panquesito7 — APPROVED** [review]
Great work, @Swastyy! Thank you for your contribution! 👍 🎉

**Review by ayaankhan98 — APPROVED** [review]

**Pull request merged**